### ISSUE TITLE
Port MobileNet models.

### ISSUE THREAD

**Issue by nsthorat** [tensorflow/tfjs-core#238]
The only vision model we've ported is SqueezeNet, however we should also port MobileNets which may be more efficient:

https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet_v1.md

We should add implementations for MobileNets in our `demos/models/` directory. If they turn out to be more efficient, we can swap it out in our demos, and host weights on GCP.

For help porting models, check out this guide:
https://deeplearnjs.org/demos/mnist/mnist.html

which uses this script to dump weights:
https://github.com/PAIR-code/deeplearnjs/blob/master/scripts/dump_checkpoint_vars.py

**Comment by chaosmail** [comment]
I just used looked at the Caffe model of MobileNet (https://github.com/shicai/MobileNet-Caffe) using the script described in #267 .

```js
fetchArrayBuffer('mobilenet.caffemodel').then(function(res){
  var model = caffe.NetParameter.decode(new Uint8Array(res));
  console.log(model.layer.map((d) => d.name + " (" + d.type + ")").join('\n'));
});
```

If all layer types are implemented, it could be loaded from Caffe.

```
data (ImageData)
label_data_1_split (Split)
conv1 (Convolution)
conv1/bn (BatchNorm)
conv1/scale (Scale)
relu1 (ReLU)
conv2_1/dw (Convolution)
conv2_1/dw/bn (BatchNorm)
conv2_1/dw/scale (Scale)
relu2_1/dw (ReLU)
conv2_1/sep (Convolution)
conv2_1/sep/bn (BatchNorm)
conv2_1/sep/scale (Scale)
relu2_1/sep (ReLU)
conv2_2/dw (Convolution)
conv2_2/dw/bn (BatchNorm)
conv2_2/dw/scale (Scale)
relu2_2/dw (ReLU)
conv2_2/sep (Convolution)
conv2_2/sep/bn (BatchNorm)
conv2_2/sep/scale (Scale)
relu2_2/sep (ReLU)
conv3_1/dw (Convolution)
conv3_1/dw/bn (BatchNorm)
conv3_1/dw/scale (Scale)
relu3_1/dw (ReLU)
conv3_1/sep (Convolution)
conv3_1/sep/bn (BatchNorm)
conv3_1/sep/scale (Scale)
relu3_1/sep (ReLU)
conv3_2/dw (Convolution)
conv3_2/dw/bn (BatchNorm)
conv3_2/dw/scale (Scale)
relu3_2/dw (ReLU)
conv3_2/sep (Convolution)
conv3_2/sep/bn (BatchNorm)
conv3_2/sep/scale (Scale)
relu3_2/sep (ReLU)
conv4_1/dw (Convolution)
conv4_1/dw/bn (BatchNorm)
conv4_1/dw/scale (Scale)
relu4_1/dw (ReLU)
conv4_1/sep (Convolution)
conv4_1/sep/bn (BatchNorm)
conv4_1/sep/scale (Scale)
relu4_1/sep (ReLU)
conv4_2/dw (Convolution)
conv4_2/dw/bn (BatchNorm)
conv4_2/dw/scale (Scale)
relu4_2/dw (ReLU)
conv4_2/sep (Convolution)
conv4_2/sep/bn (BatchNorm)
conv4_2/sep/scale (Scale)
relu4_2/sep (ReLU)
conv5_1/dw (Convolution)
conv5_1/dw/bn (BatchNorm)
conv5_1/dw/scale (Scale)
relu5_1/dw (ReLU)
conv5_1/sep (Convolution)
conv5_1/sep/bn (BatchNorm)
conv5_1/sep/scale (Scale)
relu5_1/sep (ReLU)
conv5_2/dw (Convolution)
conv5_2/dw/bn (BatchNorm)
conv5_2/dw/scale (Scale)
relu5_2/dw (ReLU)
conv5_2/sep (Convolution)
conv5_2/sep/bn (BatchNorm)
conv5_2/sep/scale (Scale)
relu5_2/sep (ReLU)
conv5_3/dw (Convolution)
conv5_3/dw/bn (BatchNorm)
conv5_3/dw/scale (Scale)
relu5_3/dw (ReLU)
conv5_3/sep (Convolution)
conv5_3/sep/bn (BatchNorm)
conv5_3/sep/scale (Scale)
relu5_3/sep (ReLU)
conv5_4/dw (Convolution)
conv5_4/dw/bn (BatchNorm)
conv5_4/dw/scale (Scale)
relu5_4/dw (ReLU)
conv5_4/sep (Convolution)
conv5_4/sep/bn (BatchNorm)
conv5_4/sep/scale (Scale)
relu5_4/sep (ReLU)
conv5_5/dw (Convolution)
conv5_5/dw/bn (BatchNorm)
conv5_5/dw/scale (Scale)
relu5_5/dw (ReLU)
conv5_5/sep (Convolution)
conv5_5/sep/bn (BatchNorm)
conv5_5/sep/scale (Scale)
relu5_5/sep (ReLU)
conv5_6/dw (Convolution)
conv5_6/dw/bn (BatchNorm)
conv5_6/dw/scale (Scale)
relu5_6/dw (ReLU)
conv5_6/sep (Convolution)
conv5_6/sep/bn (BatchNorm)
conv5_6/sep/scale (Scale)
relu5_6/sep (ReLU)
conv6/dw (Convolution)
conv6/dw/bn (BatchNorm)
conv6/dw/scale (Scale)
relu6/dw (ReLU)
conv6/sep (Convolution)
conv6/sep/bn (BatchNorm)
conv6/sep/scale (Scale)
relu6/sep (ReLU)
pool6 (Pooling)
fc7 (Convolution)
fc7_fc7_0_split (Split)
loss (SoftmaxWithLoss)
top1/acc (Accuracy)
top5/acc (Accuracy)
```

**Comment by nsthorat** [comment]
Do you know what the scale operation is?

**Comment by chaosmail** [comment]
I found this explication on SO:

**Caffe ScaleLayer**

> Computes a product of two input Blobs, with the shape of the latter Blob "broadcast" to match the shape of the former. Equivalent to tiling the latter Blob, then computing the elementwise product.
The second input may be omitted, in which case it's learned as a parameter of the layer.

Source: https://stackoverflow.com/a/37412260/5200303

It looks like an additional trainable scale operation after every BatchNorm - but I don't quite understand the purpose (I opened an issue https://github.com/shicai/MobileNet-Caffe/issues/35).

**Comment by dsmilkov** [comment]
Thanks everyone for a great discussion. Closing since we already have MobileNet ported. See [this example](https://github.com/tensorflow/tfjs-examples/tree/master/mobilenet) on the tfjs-examples repo.
